load_test.py

```diff
@@ -14,22 +14,43 @@
 import utils
 
 
-def run_main_process(concurrency, duration, dataset, dataset_q, stop_q):
+def run_main_process(rps, duration, dataset, request_q, stop_q):
     """Run the main process."""
     logging.info("Test from main process")
 
-    # Initialize the dataset_queue with 4*concurrency requests
+    start_time = time.time()
+    end_time = start_time + duration
+    if rps is not None:
+        main_loop_rps_mode(dataset, request_q, rps, start_time, end_time)
+    else:
+        main_loop_concurrency_mode(dataset, request_q, start_time, end_time)
+
+    logging.info("Timer ended, stopping processes")
+
+    # Signal users to stop sending requests
+    stop_q.put(None)
+
+    # Empty the dataset queue
+    while not request_q.empty():
+        logging.debug("Removing element from request_q")
+        request_q.get()
+
+    return
+
+def main_loop_concurrency_mode(dataset, request_q, start_time, end_time):
+    """Let all users send requests repeatedly until end_time"""
+    logging.info("Test from main process")
+
+    # Initialize the request_q with 2*concurrency requests
     for query in dataset.get_next_n_queries(2 * concurrency):
-        dataset_q.put(query)
+        request_q.put((None, query))
```

Comment:
From a clarity perspective I think it would be better to have this be a dict or object. E.g.

```suggestion
        request_q.put(dict(query=query, req_time=None))
```
or set a field on the `query` dict.

Reply:
Using a field in the query dict is much more elegant!

Reply:
Yeah, I like the idea of making it a field in the query dict.